spring-integration-core/src/main/java/org/springframework/integration/support/management/graph/GraphServer.java

```diff
@@ -0,0 +1,167 @@
+			SourcePollingChannelAdapter adapter = entry.getValue();
+			MessageSourceNode sourceNode = this.nodeFactory.sourceNode(entry.getKey(), adapter);
+			nodes.add(sourceNode);
+			MessageChannelNode channelNode = channelNodes.get(sourceNode.getOutput());
+			if (channelNode != null) {
+				links.add(new LinkNode(sourceNode.getNodeId(), channelNode.getNodeId()));
+			}
+		}
+		for (Entry<String, MessagingGatewaySupport> entry : gateways.entrySet()) {
+			MessagingGatewaySupport gateway = entry.getValue();
+			MessageGatewayNode gatewayNode = this.nodeFactory.gatewayNode(entry.getKey(), gateway);
+			nodes.add(gatewayNode);
+			MessageChannelNode channelInfo = channelNodes.get(gatewayNode.getOutput());
+			if (channelInfo != null) {
+				links.add(new LinkNode(gatewayNode.getNodeId(), channelInfo.getNodeId()));
+			}
+		}
+		for (Entry<String, MessageProducerSupport> entry : producers.entrySet()) {
+			MessageProducerSupport producer = entry.getValue();
+			MessageProducerNode producerNode = this.nodeFactory.producerNode(entry.getKey(), producer);
+			nodes.add(producerNode);
+			MessageChannelNode channelNode = channelNodes.get(producerNode.getOutput());
+			if (channelNode != null) {
+				links.add(new LinkNode(producerNode.getNodeId(), channelNode.getNodeId()));
+			}
+		}
+		for (Entry<String, IntegrationConsumer> entry : consumers.entrySet()) {
+			IntegrationConsumer consumer = entry.getValue();
+			MessageHandlerNode handlerNode = this.nodeFactory.handlerNode(entry.getKey(), consumer);
+			nodes.add(handlerNode);
+			MessageChannelNode channelNode = channelNodes.get(handlerNode.getInput());
+			if (channelNode != null) {
+				links.add(new LinkNode(channelNode.getNodeId(), handlerNode.getNodeId()));
+			}
+			if (handlerNode.getOutput() != null) {
+				channelNode = channelNodes.get(handlerNode.getOutput());
+				if (channelNode != null) {
+					links.add(new LinkNode(handlerNode.getNodeId(), channelNode.getNodeId()));
+				}
+			}
+		}
+		this.graph = new Graph(nodes, links);
+	}
+
+	/**
+	 * Rebuild the graph if new components have been added.
+	 */
+	public void rebuild() {
+		this.nodeFactory.nodeId.set(0);
+		afterSingletonsInstantiated();
+	}
+
+	private final static class NodeFactory {
+
+		private final AtomicInteger nodeId = new AtomicInteger();
+
+		private MessageChannelNode channelNode(String name, MessageChannel channel) {
+			return new MessageChannelNode(this.nodeId.incrementAndGet(), name, channel);
+		}
+
+		private MessageGatewayNode gatewayNode(String name, MessagingGatewaySupport gateway) {
+			return new MessageGatewayNode(this.nodeId.incrementAndGet(), name, gateway,
+					gateway.getRequestChannel().toString());
+		}
+
+		private MessageProducerNode producerNode(String name, MessageProducerSupport producer) {
+			MessageChannel outputChannel = producer.getOutputChannel();
+			String outputChannelName = outputChannel == null ? "__unknown__" : outputChannel.toString();
```

Comment:
I still insist that `MessageProducerSupport.getOutputChannel()` can't be `null` here already.